# Part 1. Equation of a Slime

How many late days are you using for this assignment? 0

In [ ]:
# Imports section
import pandas as pd

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import cross_val_score
from sklearn.preprocessing import PolynomialFeatures
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import StandardScaler
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.neural_network import MLPClassifier

import numpy as np


hi


## 1. Loading the dataset

In [6]:
# Using pandas load the dataset
# Output the first 15 rows of the data
# Display a summary of the table information (number of datapoints, etc.)

dataset = pd.read_csv('science_data_large.csv')

print("First 15 lines in data set:\n")
print(dataset.head(15))

print("\nInfo about data set:\n")
print(dataset.info())

First 15 lines in data set:

    Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04

Info about data set:

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL       

## 2. Splitting the dataset

In [9]:
# Take the pandas dataset and split it into our features (X) and label (y)

# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
# For grading consistency use random_state=42 

# Step 2: Define features (X) and label (y)
# Assuming the last column is the label, adjust accordingly if it’s different
X = dataset.iloc[:, :-1]  # All columns except the last one
y = dataset.iloc[:, -1]   # The last column

# Step 3: Split the dataset into a training set and a testing set (90% train, 10% test)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=42)

# Optionally, you can print the shapes of the resulting datasets
print(f"Training set shape: {X_train.shape}, {y_train.shape}")
print(f"Test set shape: {X_test.shape}, {y_test.shape}")


Training set shape: (900, 2), (900,)
Test set shape: (100, 2), (100,)


## 3. Perform a Linear Regression

In [18]:
# Use sklearn to train a model on the training set

# Create a sample datapoint and predict the output of that sample with the trained model

# Report the score for that model using the default score function property of the SKLearn model, in your own words (markdown, not code) explain what the score means

# Extract the coefficients and intercept from the model and write an equation for your h(x) using LaTeX


# Step 1: Create and train the linear regression model
model = LinearRegression()
model.fit(X_train, y_train)

# Step 2: Create a sample data point for prediction
# Here, we will create a sample point. Adjust the values as per your features.
# Replace [feature1_value, feature2_value] with actual numeric values.

# Randomly select an index from the training set
random_index = np.random.randint(0, len(X_train))  # Random index within the bounds of X_train

# Select the sample datapoint from the training set

sample_datapoint = pd.DataFrame(X_train.iloc[random_index]).T
predicted_output = model.predict(sample_datapoint)

# Step 3: Report the score of the model
score = model.score(X_test, y_test)

print(f"Predicted output for the sample data point: {predicted_output[0]}")
print(f"Model score (R²): {score}")

# Step 4: Extract coefficients and intercept
coefficients = model.coef_
intercept = model.intercept_

# Resulting equation form: y = intercept + coefficient1 * x1 + coefficient2 * x2
print(f"Coefficients: {coefficients}")
print(f"Intercept: {intercept}")

Predicted output for the sample data point: 254970.6175898721
Model score (R²): 0.8552472077276095
Coefficients: [ 866.14641337 1032.69506649]
Intercept: -409391.47958340764


Write the linear equation of a slime: (example equation: $E = mc^2$)

The model equation is given by:
$$ h(x) = -409391.47958 + 866.14641 \cdot x_1 + 1032.69507 \cdot x_2 $$


Report on score and explain meaning:

### Model Score (R²)

The R² (R-squared) score of our linear regression model is approximately **0.8552**.

### Explanation of the Score

The R² score is a statistical measure that represents the proportion of the variance for the dependent variable that is explained by the independent variables in the model. 

1. **Range of R²**: 
   R² values range from 0 to 1:
   - An R² of **1.0** indicates that the model perfectly explains the variance in the dependent variable.
   - An R² of **0.0** means that the model does not improve predictions and is equivalent to using the mean of the dependent variable as a prediction.

2. **Interpretation of 0.8552**: 
   - An R² score of **0.8552** signifies that approximately **85.52%** of the variance in the dependent variable is accounted for by the independent variables.
   - This indicates a strong fit, suggesting that the model explains a significant amount of variability in the outcome. 

## 4. Use Cross Validation

In [32]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
# For grading consistency use n_splits=5 and random_state=42

# Report on their finding and their significance

# Create a linear regression model instance
model_cv = LinearRegression()

# Step 1: Use cross_val_score to perform 5-fold cross-validation
# Setting cv=5 for grading consistency
scores = cross_val_score(model_cv, dataset.iloc[:, :-1], dataset.iloc[:, -1], cv=5, scoring='r2')

# Print the scores for each fold
print(f"Cross-validation R² scores for each fold: {scores}")
print(f"Mean cross-validation R² score: {scores.mean()}")

Cross-validation R² scores for each fold: [0.83918826 0.87051239 0.85871066 0.87202623 0.84364641]
Mean cross-validation R² score: 0.8568167899144437


Write findings here:

### Cross-Validation Findings

After performing 5-fold cross-validation on our linear regression model, we achieved the following results:

- **Cross-validation R² scores for each fold**: [0.83919, 0.87051, 0.85871, 0.87203, 0.84365]
- **Mean R² Score**: 0.85682

### Explanation of Findings

1. **Fold Scores**: The individual R² scores show how well the model performed for each of the 5 random splits of the data. Each score indicates how much of the variability in the target variable can be explained by the model for that specific fold:
   - The scores range from about **0.8392** to **0.8720**, indicating that the model performs consistently well across all the folds.

2. **Mean R² Score**: The average R² score of **0.8568** means that, on average, the model explains about **85.68%** of the variance in the dependent variable. This is a strong indication that the model is effective in making predictions.

3. **Model Performance**: Comparing this mean cross-validation score to previous scores can help us understand if the model is reliable. If this score is similar to what we saw earlier, it suggests that the model is likely not overfitting and can generalize well to new data.

4. **Overall Conclusion**: Cross-validation helps us confirm that the model works well on different parts of the data, giving us more confidence that it will perform well when applied to unseen data.

## 5. Using Polynomial Regression

In [39]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
# Perform k-fold cross validation (as above)

# Report on the metrics and output the resultant equation as you did in Part 3.

# Step 1: Augment the dataset with polynomial features of degree 2
poly = PolynomialFeatures(degree=2)
X_poly = poly.fit_transform(dataset.iloc[:, :-1])  # Create polynomial features

# Step 2: Create a new linear regression model instance
model_poly = LinearRegression()

# Step 3: Perform k-fold cross-validation
scores_poly = cross_val_score(model_poly, X_poly, dataset.iloc[:, -1], cv=5, scoring='r2')

# Print the scores for each fold
print(f"Cross-validation R² scores for polynomial regression: {scores_poly}")
print(f"Mean cross-validation R² score for polynomial regression: {scores_poly.mean()}")

# Step 4: Fit the model and extract coefficients and intercept
model_poly.fit(X_poly, dataset.iloc[:, -1])
coefficients_poly = model_poly.coef_
intercept_poly = model_poly.intercept_

# Output the coefficients and intercept
print(f"Coefficients: {coefficients_poly}")
print(f"Intercept: {intercept_poly}")

Cross-validation R² scores for polynomial regression: [1. 1. 1. 1. 1.]
Mean cross-validation R² score for polynomial regression: 1.0
Coefficients: [ 0.00000000e+00  1.20000000e+01 -1.23113604e-07 -1.05648823e-11
  2.00000000e+00  2.85714287e-02]
Intercept: 1.65727105922997e-05


Write the polynomial equation of a slime: (example equation: $E = mc^2$)

$$ h(x) = 0.000016573 + 12.000 \cdot x_1 + 0.00000000001056 \cdot x_1^2 + 2.0000 \cdot x_2^2 + 0.02857 \cdot x_2 $$

Report on the score and interpret:

### Explanation of Findings

1. **Model Performance**: An R² score of **1.0** for all folds indicates that the polynomial model perfectly explains the variance in the target variable for each fold of the cross-validation. This suggests an excellent fit, but it may also imply potential overfitting.

2. **Coefficients**:
   - The coefficients indicate the contribution of each term in the model:
     - Constant term: \(0.00001657\)
     - Coefficient for \(x_1\): \(12.0\)
     - Coefficient for squared terms and interactions are minimal to zero, affecting the overall interpretation.

3. **Intercept**: The intercept is very small, suggesting the model's predictions are centered around the origin.

# Part 2. Chronic Kidney Disease Prediction via Classification

Create code and markdown cells as needed to perform classification and report on your results

In [46]:
# Load the dataset. Then train and evaluate the classification models.

ckdset = pd.read_csv('ckd_feature_subset.csv')

print(ckdset.head())


# print(ckdset.info())
# print(ckdset.describe())

        age        bp      wbcc  appet_poor  appet_good      rbcc  Target_ckd
0  0.688312  0.333333  0.000000           1           0  0.000000           1
1  0.545455  0.333333  0.128319           1           0  0.305085           1
2  0.714286  0.500000  0.238938           1           0  0.186441           1
3  0.688312  0.333333  0.283186           0           1  0.338983           1
4  0.441558  0.333333  0.221239           1           0  0.220339           1


In [56]:
X = ckdset.drop(columns=['Target_ckd'])  # Features
y = ckdset['Target_ckd']  # Target variable


# Scale the features
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

# Create a Logistic Regression model
logistic_model = LogisticRegression(random_state=42)

# Perform 5-fold cross-validation
logistic_scores = cross_val_score(logistic_model, X_scaled, y, cv=5, scoring='accuracy')

# Print results
print(f"Logistic Regression Cross-validation Scores: {logistic_scores}")
print(f"Mean Accuracy: {logistic_scores.mean():.5f}")
print(f"Standard Deviation: {logistic_scores.std():.5f}")

Logistic Regression Cross-validation Scores: [0.96774194 0.93548387 0.90322581 0.96666667 0.93333333]
Mean Accuracy: 0.94129
Standard Deviation: 0.02404


In [57]:
# Create a Support Vector Classifier model
svc_model = SVC(random_state=42)

# Perform 5-fold cross-validation
svc_scores = cross_val_score(svc_model, X_scaled, y, cv=5, scoring='accuracy')

# Print results
print(f"SVC Cross-validation Scores: {svc_scores}")
print(f"Mean Accuracy: {svc_scores.mean():.5f}")
print(f"Standard Deviation: {svc_scores.std():.5f}")

SVC Cross-validation Scores: [0.93548387 0.96774194 0.96774194 1.         0.9       ]
Mean Accuracy: 0.95419
Standard Deviation: 0.03392


In [59]:
# Create a k-Nearest Neighbors classifier model
knn_model = KNeighborsClassifier()

# Perform 5-fold cross-validation
knn_scores = cross_val_score(knn_model, X_scaled, y, cv=5, scoring='accuracy')

# Print results
print(f"KNN Cross-validation Scores: {knn_scores}")
print(f"Mean Accuracy: {knn_scores.mean():.5f}")
print(f"Standard Deviation: {knn_scores.std():.5f}")

KNN Cross-validation Scores: [0.93548387 1.         0.90322581 1.         0.86666667]
Mean Accuracy: 0.94108
Standard Deviation: 0.05281


In [64]:
# Create a Neural Network model (default settings with one hidden layer)
nn_model = MLPClassifier(random_state=42, max_iter=1000)

# Perform 5-fold cross-validation
nn_scores = cross_val_score(nn_model, X_scaled, y, cv=5, scoring='accuracy')

# Print results
print(f"Neural Network Cross-validation Scores: {nn_scores}")
print(f"Mean Accuracy: {nn_scores.mean():.5f}")
print(f"Standard Deviation: {nn_scores.std():.5f}")

Neural Network Cross-validation Scores: [0.96774194 0.96774194 0.96774194 1.         0.93333333]
Mean Accuracy: 0.96731
Standard Deviation: 0.02109


## **Summary**

| Model                  | Mean Accuracy | Standard Deviation |
|------------------------|---------------|---------------------|
| Logistic Regression     | 0.9413        | 0.0240              |
| Support Vector Machines | 0.9542        | 0.0339              |
| k-Nearest Neighbors     | 0.94108       | 0.05281             |
| Neural Networks         | 0.96731       | 0.02109             |

### Summary of Results

After evaluating four different classification models on the CKD dataset, we find the following results:

- The **Neural Network** model achieved the highest mean accuracy of **96.73%**, demonstrating its effectiveness in classifying Chronic Kidney Disease cases based on the features provided. At first, 500 iterations was not enough to reach fully optimal convergence, and while it run properly, multiple ConvergenceWarnings were thrown. This was fixed by increasing the number of iterations from 500 to 1000.
- The **Support Vector Machine** model also performed well with a mean accuracy of **95.42%**, indicating strong predictive ability.
- The **Logistic Regression** and **k-Nearest Neighbors** models showed similar performance, both achieving mean accuracies around **94.13%** and **94.11%** respectively.

### Discussion on Performance
The neural network performed best likely due to its ability to capture complex relationships and patterns in the data through multiple layers and non-linear activation functions. The other models, while effective, might not capture such intricate interactions in feature space, particularly with a dataset that may have non-linear relationships.

Given these findings, it would be beneficial to further explore different hyperparameters or configurations for the neural network to see if additional gains in performance can be achieved.

## Results and Conclusion for Classification Experiments

In [80]:
# Experiments with Neural Network.

# Configuration 1: Single Hidden Layer with ReLU Activation

nn_model1 = MLPClassifier(hidden_layer_sizes=(10,), activation='relu', random_state=42, max_iter=2000)
nn_scores1 = cross_val_score(nn_model1, X_scaled, y, cv=5, scoring='accuracy')

# Print results
print(f"Neural Network Cross-validation Scores: {nn_scores1}")
print(f"Mean Accuracy: {nn_scores1.mean():.5f}")
print(f"Standard Deviation: {nn_scores1.std():.5f}")

Neural Network Cross-validation Scores: [0.96774194 0.96774194 0.96774194 1.         0.93333333]
Mean Accuracy: 0.96731
Standard Deviation: 0.02109


In [81]:
# Configuration 2: Two Hidden Layers with ReLU Activation

nn_model2 = MLPClassifier(hidden_layer_sizes=(50, 25), activation='relu', random_state=42, max_iter=2000)
nn_scores2 = cross_val_score(nn_model2, X_scaled, y, cv=5, scoring='accuracy')

# Print results
print(f"Neural Network Cross-validation Scores: {nn_scores2}")
print(f"Mean Accuracy: {nn_scores2.mean():.5f}")
print(f"Standard Deviation: {nn_scores2.std():.5f}")

Neural Network Cross-validation Scores: [0.96774194 0.96774194 0.96774194 1.         0.96666667]
Mean Accuracy: 0.97398
Standard Deviation: 0.01302


In [86]:
# Configuration 3: More Complex Network with ReLU Activation

nn_model3 = MLPClassifier(hidden_layer_sizes=(100, 50, 25), activation='relu', learning_rate_init=0.001, random_state=42, max_iter=1000)
nn_scores3 = cross_val_score(nn_model3, X_scaled, y, cv=5, scoring='accuracy')

# Print results
print(f"Neural Network Cross-validation Scores: {nn_scores3}")
print(f"Mean Accuracy: {nn_scores3.mean():.5f}")
print(f"Standard Deviation: {nn_scores3.std():.5f}")

Neural Network Cross-validation Scores: [0.96774194 0.96774194 0.93548387 1.         0.96666667]
Mean Accuracy: 0.96753
Standard Deviation: 0.02041


## **Summary Table**

| Configuration                          | Mean Accuracy | Standard Deviation |
|----------------------------------------|---------------|---------------------|
| Single Hidden Layer with ReLU          | 0.96731       | 0.02109             |
| Two Hidden Layers with ReLU            | 0.97398       | 0.01302             |
| More Complex Network with ReLU         | 0.96753       | 0.02041             |

## Results and Conclusion for Neural Network Experiments

### Discussion of Findings

After evaluating three different neural network configurations on the CKD dataset, the following observations can be made:

1. **Performance Comparison**:
   - **Two Hidden Layers with ReLU** achieved the highest mean accuracy of **97.40%** and the lowest standard deviation of **0.01302**, indicating not only strong performance but also consistent results across the folds.
   - The **Single Hidden Layer with ReLU** and the **More Complex Network with ReLU** configurations both performed similarly, with mean accuracies of **96.73%** and **96.75%**, respectively. The slightly higher performance of the two hidden layers demonstrates the benefit of increased model complexity.

2. **Effect of Complexity**:
   - The increase in complexity through additional hidden layers yielded better performance, likely because the model was able to learn more intricate patterns in the data. However, it is essential to balance complexity with the risk of overfitting, as indicated by the standard deviation.

3. **Activation Function**:
   - All configurations utilized the ReLU activation function, which is generally effective for deep learning tasks. No configurations with different activations were tested, but experimenting further with other activation functions (like Tanh or logistic) could provide valuable insights into their effects on performance.

### Conclusion
In conclusion, the configuration with **two hidden layers using the ReLU activation function** performed the best among those tested, indicating that this architecture may be well-suited for the given dataset. Further exploration with different architectures, activation functions, and hyperparameters could help optimize performance even further.